src/Symfony/Component/Security/Http/RateLimiter/DefaultLoginRateLimiter.php

```diff
@@ -28,21 +28,28 @@ final class DefaultLoginRateLimiter extends AbstractRequestRateLimiter
 {
     private RateLimiterFactory $globalFactory;
     private RateLimiterFactory $localFactory;
+    private ?string $secret;
 
-    public function __construct(RateLimiterFactory $globalFactory, RateLimiterFactory $localFactory)
+    public function __construct(RateLimiterFactory $globalFactory, RateLimiterFactory $localFactory, #[\SensitiveParameter] string $secret = null)
     {
+        if (!$secret) {
+            trigger_deprecation('symfony/security-http', '6.4', 'Calling "%s()" with an empty secret is deprecated, pass a non-empty string instead.', __METHOD__);
+        }
         $this->globalFactory = $globalFactory;
         $this->localFactory = $localFactory;
+        $this->secret = $secret;
     }
 
     protected function getLimiters(Request $request): array
     {
         $username = $request->attributes->get(SecurityRequestAttributes::LAST_USERNAME, '');
         $username = preg_match('//u', $username) ? mb_strtolower($username, 'UTF-8') : strtolower($username);
+        $ip = $this->secret ? hash_hmac('sha256', $request->getClientIp(), $this->secret) : $request->getClientIp();
+        $usernameAndIp = $this->secret ? hash_hmac('sha256', $username.'-'.$request->getClientIp(), $this->secret) : $username.'-'.$request->getClientIp();
```

Comment:
let's add a private method to split the hashing concern?
let's also base64-encode and truncate the hash:

```
$hash = strtr(substr(base64_encode(hash_hmac('sha256', $value, $this->secret, true)), 0, 8), '/+', '._');
```

Reply:
I added a private method as suggested.
Base64 truncated hash is a good idea.
One question though: why do you replace with `'._'` and not `'-_'` like base64 url-safe? In any case, it is url-safe as well

Reply:
no specific reasons, I borrowed this from ContainerBuilder::hash() but I'd be fine with `-` also